src/Orleans/Runtime/RequestContext.cs

```diff
@@ -92,39 +88,43 @@ public static void Set(string key, object value)
         /// <returns>Boolean <c>True</c> if the value was previously in the RequestContext key-value bag and has now been removed, otherwise returns <c>False</c>.</returns>
         public static bool Remove(string key)
         {
-            Dictionary<string, object> values = GetContextData();
+            var values = CallContextData.Value;
 
             if (values == null || values.Count == 0 || !values.ContainsKey(key))
             {
                 return false;
             }
             var newValues = new Dictionary<string, object>(values);
             bool retValue = newValues.Remove(key);
-            SetContextData(newValues);
+            CallContextData.Value = newValues;
             return retValue;
         }
 
         public static void Import(Dictionary<string, object> contextData)
         {
             if (PropagateActivityId)
             {
-                object activityIdObj;
-                if (contextData == null || !contextData.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj))
+                object activityIdObj = Guid.Empty;
+                if (contextData?.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj) != true)
                 {
-                    activityIdObj = Guid.Empty;
-                }
-
-#if NETSTANDARD
-                ActivityId.Value = (Guid) activityIdObj;
-#else
-                Trace.CorrelationManager.ActivityId = (Guid)activityIdObj;
+                    object legacyActivityIdObj = null;
+#if !NETSTANDARD
+                    if (contextData?.TryGetValue(E2_E_TRACING_LEGACY_ACTIVITY_ID_HEADER, out legacyActivityIdObj) == true)
+                    {
+                        Trace.CorrelationManager.ActivityId = (Guid) legacyActivityIdObj;
+                    }
 #endif
+                }
+                else
+                {
+                    ActivityId.Value = (Guid)activityIdObj;
```

Comment:
I strongly suspect that it will be empty in that cases, as local values ```ExecutionContext``` does not flow up to parent task as can be seen at https://github.com/dotnet/corefx/blob/b04757c307eee37dcc2b887f0b5ba08d7e866cc1/src/System.Threading/tests/AsyncLocalTests.cs#L255 and ```RequestContext``` is being imported only in tasks that is created by Orleans per request, thus ensuring that non-empty value can not end up in that place.  (+ in some places in streaming, which I didn't reviewed thoroughly, but they also seems to be safe).  But to be on completely safe side, as retrieving of empty value does not incurs much cost [value retrieving code](https://github.com/dotnet/coreclr/blob/c85c19baef5edf867c3a9a217cbe235a892d0c1e/src/mscorlib/shared/System/Threading/ExecutionContext.cs#L238), [values container class that is optimized for 
 common case of no values](https://github.com/dotnet/coreclr/blob/c85c19baef5edf867c3a9a217cbe235a892d0c1e/src/mscorlib/shared/System/Threading/AsyncLocal.cs#L128) I've just added check for non-empty async local activity id value to set it to default value.